design/thin-edge-actors-design.md

```diff
@@ -0,0 +1,282 @@
+  * The messages being serializable, they can be logged, persisted, audited, forwarded to the cloud.
+* __Testability__
+  * An actor can be tested in isolation as well as in combination.
+  * An actor can be exercised with arbitrary sequences of input messages,
+    while its output is observed and verified,
+    and some of its peers being possibly simulated.
+  * An actor's peer simulator can be as simple as a pre-registered stream of messages,
+    or as sophisticated as an error injector.
+  * In a pure actor model setting, all interactions with the system are done via actors,
+    and can be simulated, including the clock and the file system.
+
+Thin-edge leverages Rust, adding robustness to the actor model.
+
+* __Robustness__
+  * No message can be sent if not understood by the recipient.
+    The compatibility, between the messages sent to an actor
+    and the messages this actor can actually process,
+    is verified at compile-time.
+
+However, using Rust also introduce specific challenges, as we will see below.
+To name just a few:
+
+* As all the message types have to be known at compile time,
+  we need to be cautious not to make all the actor *implementations* depending on each others,
+  due to dependencies at the *message* level.
+  * The implementation of an actor that consumes telemetry data
+    must not depend on the implementations of actors producing telemetry data - and vice versa.
+* As no message can be sent without being understood by the recipient,
+  we need a shared definition of IoT messages.
+  * Consumers and producers of telemetry data must agree on what is telemetry data.
+* However, we also expect the system to be extended by independent vendors,
+  and thin-edge should not pre-defined all the messages that can be exchanged by actors.
+  * A contributor should be able to define its own set of messages usable by others.  
+
+## From actors to software components
+
+Using actors is key but not enough.
+Yes, the actor model helps thin-edge rust developers to implement components in isolation.
+However, thin-edge also aims to provide flexibility to the agent developers with ready-to-use executables,
+that can be tuned on-site for a specific use-case,
+by activating only specific features among all those available.
+
+The design must address not only the API for *actors* but also the packaging of these actors into *extensions*,
+their integration into *executables* and their *configuration* by agent developers and end-users. 
+
+* Elementary features are provided by __actors__: units of computation that interact with asynchronous messages.
+* Actors are packaged into __extensions__: rust libraries that provide implementations along interfaces,
+  input and output messages as well as instantiation services.
+* A thin-edge __executable__ makes available a collection of extensions ready to be enabled on-demand.
+* A user provided __configuration__ enables and configures sherry-picked extensions making an agent.
```

Comment:
This makes sense but I don't understand where does our current terminology "plugin" map to?

Reply:
I try to keep the term *plugin* for its current meaning: an extension written in any programming language and interacting with thin-edge over MQTT.